### PR TITLE
Event publishers to ElasticSearch and Slack

### PR DESCRIPTION
**Pull request by harshach**
### Describe your changes :

see #2117 

#
### Type of change :

- [x] New feature

#
### Frontend Preview (Screenshots) :
<p align="center">For frontend related change, please link screenshots of your changes preview! Optional for backend related changes.
</p>

#
### Checklist:

- [x] I have read the [**CONTRIBUTING**](https://docs.open-metadata.org/open-source-community/developer) document.
- [ ] I have performed a self-review of my own. 
- [ ] I have tagged my reviewers below.
- [ ] I have commented on my code, particularly in hard-to-understand areas.
- [ ] My changes generate no new warnings.
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [x] All new and existing tests passed.

#

### CODE REVIEW

File: .idea/encodings.xml [hunk 1]
```diff
@@ -5,13 +5,5 @@
...
-    <file url="file://$PROJECT_DIR$/common/src/main/resources" charset="UTF-8" />
-    <file url="file://$PROJECT_DIR$/openmetadata-dist/src/main/java" charset="UTF-8" />
-    <file url="file://$PROJECT_DIR$/openmetadata-dist/src/main/resources" charset="UTF-8" />
-    <file url="file://$PROJECT_DIR$/openmetadata-ui/src/main/java" charset="UTF-8" />
-    <file url="file://$PROJECT_DIR$/openmetadata-ui/src/main/resources/json/data" charset="UTF-8" />
-    <file url="file://$PROJECT_DIR$/openmetadata-ui/src/main/resources/ui/dist" charset="UTF-8" />
-    <file url="file://$PROJECT_DIR$/src/main/java" charset="UTF-8" />
-    <file url="file://$PROJECT_DIR$/src/main/resources" charset="UTF-8" />
   </component>
 </project>
```
Comment (sureshms) [inline .idea/encodings.xml:9]: Revert this

File: catalog-rest-service/src/main/java/org/openmetadata/catalog/CatalogApplication.java [hunk 2]
```diff
@@ -201,7 +227,6 @@ public static void main(String[] args) throws Exception {
     @Override
     public void start() throws Exception {
       LOG.info("starting the application");
-      EventPubSub.start();
```
Comment (sureshms) [inline catalog-rest-service/src/main/java/org/openmetadata/catalog/CatalogApplication.java:204]: Get rid of ManagedShutdown altogether?

File: catalog-rest-service/src/main/java/org/openmetadata/catalog/elasticsearch/ElasticSearchEventPublisher.java [hunk 3]
```diff
@@ -69,56 +68,71 @@ public void onFailure(Exception e) {
...
-    } catch (Exception e) {
-      LOG.error("failed to update ES doc", e);
     }
-    return null;
+  }
+
+  @Override
+  public void onShutdown() {
+    close();
+    LOG.info("Sample-EventPublisher-lifecycle-onShutdown");
```
Comment (sureshms) [inline catalog-rest-service/src/main/java/org/openmetadata/catalog/elasticsearch/ElasticSearchEventPublisher.java:135]: Why sample event publisher? Call it ElasticSearchEventPublisher?

**Review by sureshms — APPROVED** [review]

**Comment by sureshms** [comment]
@harshach, left some minor comments

File: catalog-rest-service/src/main/java/org/openmetadata/catalog/CatalogApplication.java (hunk 2 above)
Comment (harshach) [reply catalog-rest-service/src/main/java/org/openmetadata/catalog/CatalogApplication.java:204]: We still need to call EventPubSub.shutdown() when the server is shutting down, so keeping it for that.

**Comment by mithmatt** [comment]
Found Slack sdk which has the models which we need:
https://github.com/slackapi/java-slack-sdk/blob/main/slack-api-model/src/main/java/com/slack/api/model/Attachment.java
https://github.com/slackapi/java-slack-sdk/blob/main/slack-api-model/src/main/java/com/slack/api/model/Message.java

**Comment by harshach** [comment]
> Found Slack sdk which has the models which we need:
> https://github.com/slackapi/java-slack-sdk/blob/main/slack-api-model/src/main/java/com/slack/api/model/Attachment.java
> https://github.com/slackapi/java-slack-sdk/blob/main/slack-api-model/src/main/java/com/slack/api/model/Message.java

I wanted to avoid bringing too much of slack dependencies as in this version we are only constructing a message. But will take a look and replace the ones I have in PR

File: .gitignore [hunk 4]
```diff
@@ -42,12 +42,7 @@ logs
 *.iml
 *.ipr
 *.iws
-.idea/workspace.xml
-.idea/tasks.xml
-.idea/libraries/
-.idea/modules.xml
-.idea/dataSources.xml
-.idea/encodings.xml
+.idea/
```
Comment (mithmatt) [inline .gitignore:45]: ❤️

**Comment by harshach** [comment]
slack formatting needs to be improved but will handle it in another PR. This branch already contains multiple features

image
image
image
image

**Pull request merged**